haddock-api/src/Haddock/Interface.hs

```diff
@@ -162,17 +170,52 @@ createIfaces0 verbosity modules flags instIfaceMap =
 
 createIfaces :: Verbosity -> [Flag] -> InstIfaceMap -> ModuleGraph -> Ghc [Interface]
 createIfaces verbosity flags instIfaceMap mods = do
-  let sortedMods = flattenSCCs $ topSortModuleGraph False mods Nothing
-  out verbosity normal "Haddock coverage:"
-  (ifaces, _) <- foldM f ([], Map.empty) sortedMods
-  return (reverse ifaces)
-  where
-    f (ifaces, ifaceMap) modSummary = do
-      x <- processModule verbosity modSummary flags ifaceMap instIfaceMap
-      return $ case x of
-        Just iface -> (iface:ifaces, Map.insert (ifaceMod iface) iface ifaceMap)
-        Nothing    -> (ifaces, ifaceMap) -- Boot modules don't generate ifaces.
-
+  hsc_env <- getSession
+  let dflags = hsc_dflags hsc_env
+  let full_mg = topSortModuleGraph False mods Nothing
+
+  graph' <- liftIO $ forM full_mg $ \m ->
+    case m of
+      CyclicSCC _ -> return Nothing
+      AcyclicSCC ms -> do
+        mvar <- newEmptyMVar
+        return (Just (moduleName $ ms_mod ms, (ms, mvar)))
+  let graph = catMaybes graph'
+      dep_graph = Map.fromList graph
+
+  n_jobs <- case parMakeCount dflags of
+                   Nothing -> liftIO getNumProcessors
+                   Just n -> return n
+
+  par_sem <- liftIO $ newQSem n_jobs
+
+  let
+    withSem sem = bracket_ (waitQSem sem) (signalQSem sem)
+
+    spawnWorkers s = forM graph $ \(_mn, (m, mvar)) ->
+      forkIOWithUnmask $ \_unmask -> withSem par_sem $ do
```

Comment:
I think you want the `withSem` lower down, right before the `processModule`. Otherwise you can get a deadlock, where some modules have grabbed resources from the semaphore and are blocked on dependencies, but the compilation of the dependencies can't proceed because they need the semaphore.